README.md
# runtainer

Run anything as a container (in local Kubernetes cluster).

- [runtainer](#runtainer)
  - [Getting Started](#getting-started)
    - [Prerequisites](#prerequisites)
      - [Mac](#mac)
      - [Windows Native](#windows-native)
      - [Windows WSL2](#windows-wsl2)
      - [Linux](#linux)
    - [Installation](#installation)
      - [Brew](#brew)
      - [From Release](#from-release)
      - [From Sources](#from-sources)
    - [Usage](#usage)
      - [Basic container run](#basic-container-run)
      - [Exec into container and stay there](#exec-into-container-and-stay-there)
      - [Run As](#run-as)
      - [Extra volumes and port forwarding](#extra-volumes-and-port-forwarding)
      - [Extra ENV variables](#extra-env-variables)
      - [Injecting secrets](#injecting-secrets)
      - [Custom directory](#custom-directory)
      - [Piping](#piping)
      - [Private Images](#private-images)
      - [Disable automatic discovery](#disable-automatic-discovery)
      - [Troubleshooting](#troubleshooting)
    - [Configuration](#configuration)
  - [Why](#why)
  - [Disclaimer](#disclaimer)
  - [How](#how)

## Getting Started

### Prerequisites

You should be running local Kubernetes cluster. Your Kube Config (`~/.kube/config` or `KUBECONFIG`) should be pointing to that local cluster. RT will use your default context and a namespace set in it.

RT will run just fine if you point it to a remote cluster - but it would make a very little sense since environment variables and local paths would not exist on the remote cluster node. Most likely the pod will just fail to start.

Status for some of the possible setups that were tested:

#### Mac

- Docker for Desktop: :white_check_mark:
- Lima+K3s: :white_check_mark:
- Lima+K3d: :grey_question:
- Rancher Desktop (K3s): :white_check_mark:
- Rancher Desktop (K3d): :grey_question:

#### Windows Native

- Docker for Desktop: :x: (it would work but Docker VM uses `/mnt/host` instead of just `/mnt` - RT will have to figure out detecting that automatically - it can be fixed)
- WSL2+K3s: :grey_question: (seems like k3s required systemd which is missing in Ubuntu WSL2 - it should be possible, but I didn't bother to make it work and test)
- WSL2+K3d: :grey_question: (seems like k3s required systemd which is missing in Ubuntu WSL2 - it should be possible, but I didn't bother to make it work and test)
- Rancher Desktop (K3s): :white_check_mark:
- Rancher Desktop (K3d): :grey_question:

#### Windows WSL2

Please note VMs in WSL can't access each other directly - they are isolated. If your kubernetes cluster is in different VM than your code in (which is the case for Docker for Desktop as well as for Rancher Desktop) - this will not work with RT. Luckily you can work around that by mounting your code location via `/mnt/wsl` path witch is a `tmpfs` shared across all VMs. Please note this location is ephemeral - all contents will be nuked after restarts. So we will be using it to share a mount point, so the actual content stays on one of the VMs.

On the VM with code:

```ps1
wsl mkdir -p /mnt/wsl/repos
wsl sudo mount --bind ~/repos /mnt/wsl/repos
```

On the Kubernetes VM:

```ps1
wsl -d rancher-desktop mkdir -p ~/repos
wsl -d rancher-desktop mount --bind /mnt/wsl/repos ~/repos
```

As the mount point `/mnt/wsl` is getting nuked every reboot - adding that to `/etc/fstab` wouldn't work. You have to repeat that after every reboot. Also - this doesn't taking care of automatically mounting other locations of interest such as `~/.aws` and `~/.kube`, nor does the Kubernetes VM will know about environment variables in the RT VM. So the entire setup while theoretically possible - hugely cumbersome.

- Docker for Desktop: :grey_question: (see disclaimer above)
- WSL2+K3s: :grey_question: (seems like k3s required systemd which is missing in Ubuntu WSL2 - it should be possible, but I didn't bother to make it work and test)
- WSL2+K3d: :grey_question: (seems like k3s required systemd which is missing in Ubuntu WSL2 - it should be possible, but I didn't bother to make it work and test)
- Rancher Desktop (K3s): :grey_question: (see dislaimer above)
- Rancher Desktop (K3d): :grey_question: (see dislaimer above)

#### Linux

- Docker CE: :grey_question:
- K3s: :white_check_mark:
- K3d: :grey_question:
- Rancher Desktop (K3s): :grey_question:
- Rancher Desktop (K3d): :grey_question:

### Installation

#### Brew

Installation with `brew` is coming later.

#### From Release

Download a binary from [Releases](https://github.com/plumber-cd/runtainer/releases). All binaries are built with GitHub Actions and you can inspect [how](.github/workflows/release.yml).

Don't forget to add it to your `PATH`.

#### From Sources

You can build it yourself, of course (and Go made it really easy):

```bash
go install github.com/plumber-cd/runtainer@${version}
```

Don't forget to add it to your `PATH`.

### Usage

See full CLI [docs](./runtainer.md).

Image name serves as a separator. To its left RT's own arguments are expected. To it's right - command and arguments to be passed to the container.

Command separated from the arguments by a `--`, as described in the POSIX chapter 12.02, Guideline 10: https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html#tag_12_02. Both command and arguments are two arrays of strings in Kubernetes API, so from the user perspective - separator is not required to be used explicitly as K8s will accept it just fine without it. You will want to use a separator when you need to use default `ENTRYPOINT` (`.spec.container[].command`) from the image while passing custom arguments `CMD` (`.spec.container[].args`) to it.

There are 3 modes RT can run an image:

1. By default, preferred method is `PodRunModeModeExec`. RT will use `.spec.container[].command = ["cat"]` without arguments to start a pod (which will keep it running doing nothing indefinitely) and then it will exec into that pod executing combined command and arguments from the RT input. Technically, in this mode there is no difference between interactive and non-interactive mode, so `--interactive` flag is effectively ignored.

        runtainer alpine sh # interactive
        runtainer alpine whoami # non-interactive

   This mode is preferred because RT can guarantee nothing will get executed in the container that the user can miss in the terminal and because this mode does not have downsides of other modes.

2. When there is no command passed to the RT (even if arguments were passed) - RT will not be able to determine default `ENTRYPOINT` of the image to run the previous mode. In this case RT will use `PodRunModeModeAttach` mode - it will set `.spec.container[].args` accordingly to RT input and start a pod and then attach to it.

        runtainer alpine -- sh

   In this mode there's always a chance that the user may miss some activity in the container from the moment the pod became `running` and RT was able to attach to it.
   Probably due to that (or might be some other weird bug in `k8s.io/client-go`) - you almost 100% guaranteed not going to see initial command prompt after the attachment, that's why you will be presented with a message `If you don't see a command prompt, try pressing enter.` - the same one `kubectl run` will present you with.
   Also, this mode requires user to explicitly tell RT if that is interactive session or not. Otherwise if it is not interactive - by the time RT will try to attach the pod might already be in `succeeded` state, so RT will fail with an ugly message. You can try it: `runtainer alpine -- whoami`.
   Lastly, regardless of `--stdin` value - RT will not be able to pass its own StdIn to the container. The main container process has been already started by kubelet without RT participation and RT attaches to it mid-term. Even though any input after the attachment will be transferred to the container just fine - any initial StdIn such as pipe from another program will not be possible:

        echo hi | runtainer alpine -- xargs echo # will not print hi

3. Third mode `PodRunModeModeLogs` essentially is complimentary to `PodRunModeModeAttach` - it's basically the same one but it's expecting pod to become `succeeded` without user interaction so it will not ever try to attach to it. Instead it will use pod logs API to stream logs which can be done as long as the pod exists (even after it succeeded before it's being garbage-collected).

        runtainer --interactive=false alpine -- whoami

   One big downside of this mode is that it essentially disables both `--stdin` and `--tty` - all it does is just streaming logs in read-only mode.
   It also disables `--port` for the same reason.

With that - a simple rule is to try to use default `PodRunModeModeExec` mode as much as possible, unless `ENTRYPOINT` of the image is not known - then one of the `PodRunModeModeAttach` or `PodRunModeModeLogs` might help.

Below is a few more examples:

#### Basic container run

```bash
runtainer alpine whoami
runtainer maven:3.6.3-jdk-14 mvn -version
runtainer maven:3.6.3-jdk-14 mvn -- -version
```

#### Exec into container and stay there

```bash
runtainer alpine sh
```

#### Run As

By default, RT will use `runAsUser` and `runAsGroup` for security context of the pod to make it UID and GID of the local host user. This might not always work and it depends on the image and what you are trying to do with it.

Some images and software in them might rely on the user home directory. As RT will essentially change the `id` of the user - the home directory bundled with the image will no longer be accessible.

In other cases, some software might not like that the current `id` is set to the user that doesn't exists in the system. For example a simple `whoami` will not work:

```bash
runtainer -q alpine whoami
whoami: unknown uid 1000
```

#### Extra volumes and port forwarding

This one is especially fun as basically this Jenkins Master in the container will have aws/k8s/etc access right from your laptop. How cool is that, huh?

```bash
JENKINS_VERSION=2.289.1-lts
runtainer -v $(cd && pwd)/.jenkins:/var/jenkins_home -p 8080:8080 jenkins/jenkins:${JENKINS_VERSION}
```

Or maybe you want to test a Jenkins plugin?

```bash
runtainer -p 8080:8080 maven:3.8.1-jdk-8 mvn -Dhost=0.0.0.0 clean hpi:run
```

#### Extra ENV variables

```bash
runtainer -e PROFILE maven:3.6.3-jdk-14 mvn -version
# OR
runtainer -e PROFILE=foo maven:3.6.3-jdk-14 mvn -version
```

It will also automatically discover ENV variables with `RT_VAR_*` and `RT_EVAR_*` prefixes.

```bash
RT_VAR_FOO=foo RT_EVAR_BAR=bar runtainer alpine env
```

You will see that `RT_VAR_FOO` was passed to the container as-is, and `RT_EVAR_BAR` was passed to the container as `BAR` i.e. removing the prefix.

#### Injecting secrets

See [example](examples/secrets).

#### Custom directory

```bash
runtainer -d /tmp alpine touch hi.txt
```

#### Piping

```bash
echo hi | runtainer -t=false alpine xargs echo
```

By default `runtainer` run containers with `--stdin` and `--tty`, so you could use container interactively with minimum extra moves.
But for piping we need to explicitly disable `--tty` so it could read stdin from a pipe and not a virtual terminal.
You can even pipe multiple `runtainer` calls too:

```bash
runtainer alpine echo hi | runtainer -t=false alpine xargs echo
```

#### Private Images

First, you will have to create an Image Pull Secret in whichever way you normally prefer, for example:

```bash
kubectl create secret docker-registry my-registry \
    --docker-server=<registry> \
    --docker-username=<username> \
    --docker-password=<pwd> \
    --docker-email=<email>
```

Now you all you need is to tell RT the secret name to use:

```bash
runtainer --secret my-registry registry/image
```

#### Disable automatic discovery

You can optionally disable unwanted automatic discovery or its parts. See [example](examples/disable-discovery).

#### Troubleshooting

Use `--log` to make it write additional diag messages to a log file in the current working directory. Use `--debug` to write even more verbose diag messages.

### Configuration

The tool can be configured with both files and ENV variables.

ENV variables prefixed with `RT_` may be used, i.e. `RT_LOG=true`.

For config files it supports many config file formats, including JSON, TOML, YAML, HCL, envfile and Java properties formats.
Log files can be located in multiple locations, and they read on top of each other with merging in the following order:

1. Global system-level config in `~/.runtainer/config` (can be overridden with `--config`)
1. Local config file `.runtainer.yaml` (or any other supported extension) in the directory `runtainer` is started in
1. Local config file `.runtainer.yaml` (or any other supported extension) in the directory `runtainer` specified by `--dir` (if any)

This can be useful to configure on per-project or per-directory basis with committing that config to Git.

Config files and env variables apart from being an alternative source for CLI args configuration, can also be used to configure various host, image or container related things. See the following yaml example:

```yaml
log: true
environment:
  # Var FOO have an explicit hardcoded value here
  FOO: foo
  # Var BAR does not have a value, so it will be mirrored from the host at runtime
  BAR: null
volumes:
  hostMapping:
    - src: /home/foo/bar
      dest: /root/baz
```

Run it with `--debug` and you will be able to visually inspect what the tool automatically discovers for you.

## Why

Container's purpose is to run every process in its own isolated, self-contained and portable environment.
This is a very well known and widely adopted concept in application development.
What is less commonly used is the fact that you can also pack your tools as a container image and run it anywhere. Think of container images like a `brew` or `chocolatey` or even `pyenv`. You never actually have to install anything on your workstation anymore. You can easily choose which version of what you want to use by just changing the image tag.

Consider the following example.
You need to run AWS CLI.
You either pack it as a container image yourself or use some existing image, like [`amazon/aws-cli`](https://hub.docker.com/r/amazon/aws-cli).
You can do something like this:

```bash
docker run amazon/aws-cli sts get-caller-identity
```

Of course, isolated container will have no AWS credentials whatsoever, but if you have an active session on your workstation, you can do something like:

```bash
docker run -v $(cd && pwd)/.aws:/root/.aws -e AWS_PROFILE -e AWS_DEFAULT_REGION amazon/aws-cli sts get-caller-identity
```

You can go very far with this concept. You can mount other similar known locations (let's say `~/.m2`, `~/.kube` etc.), current working directory, and pass any known environment variables to the inside. Basically your workstation considered stateful and used as a persistent storage, while various containers can easily reproduce any runtime environment you need (any os/tools/etc).

You can also easily choose which version of your tool you want to run at any given time:

```bash
docker run amazon/aws-cli:2.0.40 sts get-caller-identity
```

And in case you are using multiple versions all the time, which is common especially in a micro-stack world for stability purposes, you don't need to re-install the software all the time. So, containers also can be used as `pyenv`, `virtualenv` or `rvm` of some sort.

Also, part of the image could be any number of useful wrappers and scripts, so you can package complex routines involving multiple tools into one executable artifact and distribute it to your team, so they can run it 1:1 as you did.

As you probably already imagined, the only problem with this approach is - a lot of typing. You got to provide so many arguments to mount all the locations and pass in all the environment variables. This is a lot of routine work and it is boring.

This tool basically aims to remove that overhead. It will automatically discover all known file system locations and environment variables and pass them to the container so you don't have to. It is basically a smart equivalent for `docker run` or `kubernetes run` commands.

For portability under the hood RT is using Kubernetes. That way no matter your container engine choice - the tool will always be able to run.

## Disclaimer

Be mindful of what you run.
This tool's mere purpose is to run the container as if its internals were installed on your host.
That means the usual isolation attributed to the containers in general is compromised, any potential malicious programs could easily escape to the host level and mess with your stuff.
You should really know and trust containers you run with RunTainer as like you were to do it with regular software you install on your desktop.

## How

This tool will do the following:

1. Discover and mount to your container well known locations (such as `~/.ssh`, `~/.aws` etc) as well as user-defined locations
1. Expose to your container well known environment variables (such as `AWS_PROFILE`, `AWS_DEFAULT_REGION` etc) as well as user-defined environment variables
1. Mount `cwd` and make it a `cwd` inside the container
1. Make UID/GID to match your host inside the container
1. Run the command you specified in your container (or default `ENTRYPOINT`)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.0] - 2022-10-12

### BREAKING CHANGES

- Will no longer set `fsGroup` to the current host GID by default.
- Will now set `runAsUser` to the current host UID and `runAsGroup` to the current host GID instead.
- Added new `--run-as-current-user` and `--run-as-current-group` options to disable this new behavior.

Should be fine for most cases, but note that depending on what is in the image - it might break stuff sometimes.
Some images might rely on the user home directory. This new behavior is changing the user `id`, so the home directory bundled with the image no longer accessible.
In other cases, some software might not like that the current `id` set to the user that doesn't exists in the system. I.e.:

```bash
runtainer -q alpine whoami
whoami: unknown uid 1000
```

### Also in this release

- Updated all dependencies

## [0.1.6] - 2022-04-30

### Changed

- Force secure `0600` on the mounted secrets, but since we are using `fsGroup` - kubernetes will force `0640` to it.
- Fix error adding multiple `items` to `--secret-volume` by allowing them to be added individually as `item`.
  This usually warrants major version bump since it is not backward compatible, but it is a hotfix to a feature that was just released and never worked, so it's fine.
- Implement `--disable-discovery` that allows to disable all or one-by-one elements of automatic discovery

## [0.1.5] - 2022-04-30

### Added

- Add options to `--secret-env` and `--secret-volume`
- With `--secret-env` you can now specify custom `prefix`
- With `--secret-volume` you can now specify custom `mountPath` and `items`

## [0.1.4] - 2022-04-30

### Added

- Forward `$SSH_AUTH_SOCK` in to the container
- Implement secrets injection `--secret-env` and `--secret-volume`

## [0.1.3] - 2021-10-06

- Implemented `--secret`/`-S`

## [0.1.2] - 2021-09-28

- Fix terminal TTY (regression in `v0.1.1`)

## [0.1.1] - 2021-09-27

- Fixed terminal resizing and wrapping

## [0.1.0] - 2021-09-26

### BREAKING CHANGES

- Docker CLI is no longer used
- RT now uses K8s Go Client to run container in a pod
- There are substantial breaking changes in RT CLI arguments to back up that change

## [0.0.2] - 2020-08-25

### Added

- Added `--dry-run` mode to print out what it normally would run otherwise
- Implemented env variables discovery mechanism similar to what's for volumes
- `env` was moved out of `host` in config files - now it's a root level map
- [Discovery/AWS]: Add `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, `AWS_SESSION_TOKEN`, `AWS_ROLE_SESSION_NAME`, `AWS_STS_REGIONAL_ENDPOINTS` and `AWS_SDK_LOAD_CONFIG`
- Helm discovery was not enabled
- Implemented Terraform discovery

## [0.0.1] - 2020-08-22

### Added

- Initial release


cmd/root.go
package cmd

import (
	"encoding/json"
	llog "log"
	"os"
	"path/filepath"
	"strings"

	homedir "github.com/mitchellh/go-homedir"
	"github.com/plumber-cd/runtainer/backends/k8s"
	"github.com/plumber-cd/runtainer/log"
	"github.com/plumber-cd/runtainer/utils"
	"github.com/spf13/cobra"
	"github.com/spf13/viper"
)

var (
	cfgFile string

	rootCmd = &cobra.Command{
		Use:                   "runtainer [runtainer flags] image [container cmd] [-- [container args]]",
		Short:                 "Run anything as a Container",
		Long:                  "See https://github.com/plumber-cd/runtainer/README.md for details",
		DisableFlagsInUseLine: true,
		Args:                  cobra.MinimumNArgs(1),
		Run: func(cmd *cobra.Command, args []string) {
			log.Debug.Print("Start root command execution")

			// the args will contain all the args unrecognized by cobra after the first positional arg (not dash prefixed)
			// the first not dash prefixed arg must be the image name
			imageName := args[0]
			log.Debug.Printf("Image: %s", imageName)

			// rest of the args split by -- delimiter
			// See POSIX chapter 12.02, Guideline 10: https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html#tag_12_02
			// On the left, args considered to be passed to the backend (docker/kubectl/etc), on the right args considered to be passed to the container
			containerCmd, containerArgs := splitArgs(args[1:])

			// run discovery routines that will publish all the facts to viper for backend engine to interpret
			discover(imageName)

			// just for debugging, dump full viper data before passing it to the backends
			allSettings, err := json.MarshalIndent(viper.AllSettings(), "", "  ")
			if err != nil {
				log.Normal.Panic(err)
			}
			log.Debug.Printf("Settings: %s", string(allSettings))

			k8s.Run(containerCmd, containerArgs)
		},
	}
)

// Execute executes the root command.
func Execute() error {
	return rootCmd.Execute()
}

func init() {
	cobra.OnInitialize(initConfig)

	rootCmd.PersistentFlags().StringVarP(&cfgFile, "config", "c", "", "global config file (default is $HOME/.runtainer.yaml)")

	rootCmd.PersistentFlags().BoolP("quiet", "q", false, `Enable quiet mode.
	By default runtainer never prints to StdOut,
	reserving that channel exclusively to the container.
	But it does print messages to StdErr.
	Enabling quiet mode will redirect all messages to the info logger.
	If --log mode was not enabled - these messages will be discarded.`)
	if err := viper.BindPFlag("quiet", rootCmd.PersistentFlags().Lookup("quiet")); err != nil {
		llog.Panic(err)
	}

	rootCmd.PersistentFlags().Bool("log", false, "Enables info logs to file")
	if err := viper.BindPFlag("log", rootCmd.PersistentFlags().Lookup("log")); err != nil {
		llog.Panic(err)
	}

	rootCmd.PersistentFlags().Bool("debug", false, "Enables info and debug logs to file")
	if err := viper.BindPFlag("debug", rootCmd.PersistentFlags().Lookup("debug")); err != nil {
		llog.Panic(err)
	}

	rootCmd.PersistentFlags().BoolP("interactive", "i", true, `Disable to not to attach to the container.
	By default we wait till pod becomes Running and then - attaching to it.
	If container expected to run a script in non-interactive mode and exit,
	- the tool might try to attach to the container that is already finished and fail.
	Disable interactive mode in this case - then it will not attempt to attach
	and instead will just stream logs until containe becomes either Succeeded or Failed.
	This automatically disables --stdin and --tty.`)
	if err := viper.BindPFlag("interactive", rootCmd.PersistentFlags().Lookup("interactive")); err != nil {
		llog.Panic(err)
	}

	rootCmd.PersistentFlags().BoolP("stdin", "s", true, "Redirect host StdIn to the container")
	if err := viper.BindPFlag("stdin", rootCmd.PersistentFlags().Lookup("stdin")); err != nil {
		llog.Panic(err)
	}

	rootCmd.PersistentFlags().BoolP("tty", "t", true, "Enable TTY, disable if piping something to stdin")
	if err := viper.BindPFlag("tty", rootCmd.PersistentFlags().Lookup("tty")); err != nil {
		llog.Panic(err)
	}

	rootCmd.PersistentFlags().StringP("dir", "d", "", "Use different folder to make a CWD in the container (default is the host CWD)")
	if err := viper.BindPFlag("dir", rootCmd.PersistentFlags().Lookup("dir")); err != nil {
		llog.Panic(err)
	}

	rootCmd.PersistentFlags().StringSliceP("env", "e", []string{}, "Mapping for env, i.e. --env AWS_PROFILE or --env AWS_PROFILE=foo")
	if err := viper.BindPFlag("env", rootCmd.PersistentFlags().Lookup("env")); err != nil {
		llog.Panic(err)
	}

	rootCmd.PersistentFlags().StringSlice("secret-env", []string{}, "Mapping for env secrets, i.e. --secret-env foo secret-env bar")
	if err := viper.BindPFlag("secrets.env", rootCmd.PersistentFlags().Lookup("secret-env")); err != nil {
		llog.Panic(err)
	}

	rootCmd.PersistentFlags().StringSliceP("port", "p", []string{}, "Mapping for ports, i.e. --port 8080:8080")
	if err := viper.BindPFlag("port", rootCmd.PersistentFlags().Lookup("port")); err != nil {
		llog.Panic(err)
	}

	rootCmd.PersistentFlags().StringSliceP("volume", "v", []string{}, "Mapping for volumes, i.e. --volume /data:/data")
	if err := viper.BindPFlag("volume", rootCmd.PersistentFlags().Lookup("volume")); err != nil {
		llog.Panic(err)
	}

	rootCmd.PersistentFlags().StringSlice("secret-volume", []string{}, "Mapping for env secrets, i.e. --secret-volume foo secret-volume bar")
	if err := viper.BindPFlag("secrets.volumes", rootCmd.PersistentFlags().Lookup("secret-volume")); err != nil {
		llog.Panic(err)
	}

	rootCmd.PersistentFlags().StringP("secret", "S", "", "Optionally, provide a name of the secret to be used for the image pull")
	if err := viper.BindPFlag("secret", rootCmd.PersistentFlags().Lookup("secret")); err != nil {
		llog.Panic(err)
	}

	rootCmd.PersistentFlags().StringSlice("disable-discovery", []string{}, "Disable individual discovery mechanisms")
	if err := viper.BindPFlag("discovery.disabled", rootCmd.PersistentFlags().Lookup("disable-discovery")); err != nil {
		llog.Panic(err)
	}

	rootCmd.PersistentFlags().BoolP("run-as-current-user", "U", true, "Will set runAsUser to the current host UID.")
	if err := viper.BindPFlag("run-as-current-user", rootCmd.PersistentFlags().Lookup("run-as-current-user")); err != nil {
		llog.Panic(err)
	}

	rootCmd.PersistentFlags().BoolP("run-as-current-group", "G", true, "Will set runAsGroup to the current host GID. Ignored if -U=false. If disabled - will set fsGroup to the current host GID instead.")
	if err := viper.BindPFlag("run-as-current-group", rootCmd.PersistentFlags().Lookup("run-as-current-group")); err != nil {
		llog.Panic(err)
	}

	rootCmd.PersistentFlags().Bool("dry-run", false, "Dry Run mode will not execute the container, only print to StdOut a pod spec it would have run.")
	if err := viper.BindPFlag("dry-run", rootCmd.PersistentFlags().Lookup("dry-run")); err != nil {
		llog.Panic(err)
	}

	rootCmd.Flags().SetInterspersed(false)
}

func initConfig() {
	log.Debug.Print("Read viper configs")

	if cfgFile != "" {
		log.Debug.Printf("Using custom global config path: %s", cfgFile)

		exists, err := utils.FileExists(cfgFile)
		if err != nil {
			log.Normal.Panic(err)
		}
		if !exists {
			log.Normal.Fatalf("Global config file not found: %s", cfgFile)
		}

		// Use config file from the flag.
		viper.SetConfigFile(cfgFile)
	} else {
		log.Debug.Print("Using default global config in user home")

		// Find home directory.
		home, err := homedir.Dir()
		if err != nil {
			log.Error.Panic(err)
		}

		// Search config in home directory with name ".runtainer" (without extension).
		viper.SetConfigName("config")
		viper.AddConfigPath(filepath.Join(home, ".runtainer"))
	}

	log.Debug.Print("Enabling ENV parsing for viper")
	// This is so we can set any nested viper settings via env variables, replacing every . with _
	viper.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
	// RT short from RunTainer
	viper.SetEnvPrefix("RT")
	viper.AutomaticEnv()

	if err := viper.ReadInConfig(); err != nil {
		switch err.(type) {
		case viper.ConfigFileNotFoundError:
			log.Debug.Printf("Global %s, skipping...", err)
		default:
			log.Error.Panic(err)
		}
	} else {
		log.Debug.Print("Using global config file:", viper.ConfigFileUsed())
	}

	// try to read (if exists) local config file in the cwd
	cwd, err := os.Getwd()
	if err != nil {
		log.Error.Panic(err)
	}
	readLocalConfig(cwd)

	// by the time we load viper configs, we haven't discovered the host yet,
	// so we couldn't use host.Cwd above
	// but just in case if user provided some custom directory, check for possible config there too
	if d := viper.GetString("dir"); d != "" {
		readLocalConfig(d)
	}

	log.Debug.Print("Viper configs loaded, re-initialize logger in case anything changed...")
	log.SetupLog()
}

// readLocalConfig read viper config in the directory.
// Due to https://github.com/spf13/viper/issues/181,
// seems like there's not really a way to override with multiple config files.
// So we will read local config files into a separate viper instances, and then use MergeConfigMap with AllSettings.
func readLocalConfig(d string) {
	log.Debug.Printf("Reading config file in %s", d)

	v := viper.New()
	v.SetConfigName(".runtainer")
	v.AddConfigPath(d)
	if err := v.ReadInConfig(); err != nil {
		switch err.(type) {
		case viper.ConfigFileNotFoundError:
			log.Debug.Printf("Local %s, skipping...", err)
		default:
			log.Error.Panic(err)
		}
	} else {
		log.Debug.Print("Using local config file:", v.ConfigFileUsed())
		if err := viper.MergeConfigMap(v.AllSettings()); err != nil {
			log.Error.Panic(err)
		}
	}
}

// splitArgs as per that POSIX standard, find the -- delimiter and split args by it
func splitArgs(args []string) ([]string, []string) {
	log.Debug.Printf("args: %s", strings.Join(args, " "))
	backendArgs := args
	var containerArgs []string
	for i := range args {
		if args[i] == "--" {
			backendArgs = args[:i]
			containerArgs = args[(i + 1):]
			break
		}
	}
	log.Debug.Printf("backendArgs: %s", strings.Join(backendArgs, " "))
	log.Debug.Printf("containerArgs: %s", strings.Join(containerArgs, " "))
	return backendArgs, containerArgs
}
